common/flatpak-run-wayland.c

```diff
@@ -0,0 +1,218 @@
+    return FALSE;
+
+  registry = wl_display_get_registry (display);
+  wl_registry_add_listener (registry, &registry_listener,
+                            &security_context_manager);
+  ret = wl_display_roundtrip (display);
+  wl_registry_destroy (registry);
+  if (ret < 0)
+    goto out;
+
+  if (!security_context_manager) {
+    *available_out = FALSE;
+    goto out;
+  }
+
+  socket_path = create_wl_socket ("wayland-XXXXXX");
+  if (!socket_path)
+    goto out;
+
+  unlink (socket_path);
+
+  listen_fd = socket (AF_UNIX, SOCK_STREAM, 0);
+  if (listen_fd < 0)
+    goto out;
+
+  sockaddr.sun_family = AF_UNIX;
+  snprintf (sockaddr.sun_path, sizeof(sockaddr.sun_path), "%s", socket_path);
+  if (bind (listen_fd, (struct sockaddr *) &sockaddr, sizeof (sockaddr)) != 0)
+    goto out;
+
+  if (listen (listen_fd, 0) != 0)
+    goto out;
+
+  sync_fd = flatpak_bwrap_add_sync_fd (bwrap);
+  if (sync_fd < 0)
+    goto out;
+
+  security_context = wp_security_context_manager_v1_create_listener (security_context_manager,
+                                                                     listen_fd,
+                                                                     sync_fd);
+  wp_security_context_v1_set_sandbox_engine (security_context, "flatpak");
+  wp_security_context_v1_set_app_id (security_context, app_id);
+  wp_security_context_v1_set_instance_id (security_context, instance_id);
+  wp_security_context_v1_commit (security_context);
+  wp_security_context_v1_destroy (security_context);
+  if (wl_display_roundtrip (display) < 0)
+    goto out;
+
+  flatpak_bwrap_add_args (bwrap,
+                          "--ro-bind", socket_path, "/run/flatpak/wayland-0",
+                          NULL);
+  flatpak_bwrap_set_env (bwrap, "WAYLAND_DISPLAY", "/run/flatpak/wayland-0", TRUE);
+
+  res = TRUE;
+
+out:
+  if (listen_fd >= 0)
+    close (listen_fd);
+  if (security_context_manager)
+    wp_security_context_manager_v1_destroy (security_context_manager);
+  wl_display_disconnect (display);
+  return res;
+}
+
+#endif /* ENABLE_WAYLAND_SECURITY_CONTEXT */
+
+/**
+ * flatpak_run_add_wayland_args:
+ *
+ * Returns: %TRUE if a Wayland socket was found.
+ */
+gboolean
+flatpak_run_add_wayland_args (FlatpakBwrap *bwrap,
+                              const char   *app_id,
+                              const char   *instance_id)
+{
+  const char *wayland_display;
+  g_autofree char *user_runtime_dir = flatpak_get_real_xdg_runtime_dir ();
+  g_autofree char *wayland_socket = NULL;
+  g_autofree char *sandbox_wayland_socket = NULL;
+  gboolean res = FALSE;
+  struct stat statbuf;
+
+#ifdef ENABLE_WAYLAND_SECURITY_CONTEXT
+  gboolean security_context_available = FALSE;
+  if (flatpak_run_add_wayland_security_context_args (bwrap, app_id, instance_id,
+                                                     &security_context_available))
+    return TRUE;
+  if (security_context_available)
+    return FALSE;
+#endif /* ENABLE_WAYLAND_SECURITY_CONTEXT */
```

Comment:
I'd appreciate a bit of an explanation on what's going on here; the `if` check seems a bit backwards. AIUI, this would mean not running with Wayland at all, thus running *on X11*; why is that desired? You'd be breaking things like fractional scaling and mixed monitor scales when doing so. And even a Wayland socket *without* a security_context is leaps better than an X11 socket on its own.

Reply:
I'm still not sure on *why* we'd not prefer to use Wayland. What if a compositor doesn't support the protocol (like an old Mutter) but doesn't provide any iffy protocol extensions? Then this would mean all Flatpak apps run on X11 (if Flatpak is built with security_context support), which would result in a degraded UX. 

It seems more appropriate to log the issue and continue on.

Reply:
If the compositor doesn't support the protocol, the current logic will just set up a regular Wayland connection. No X11 used.

Reply:
It seems the general flow of the code was confusing me. There's `flatpak_run_add_wayland_security_context_args`, which does all of the security_context *and* wayland socket setup, and `flatpak_run_add_wayland_args`, which does the normal wayland stuff.

`flatpak_run_add_wayland_args` calls `flatpak_run_add_wayland_security_context_args` to completely set up the socket from here. If the protocol is available but failed to be set up, it'll bail out, running on X11. I would still prefer that to instead log an error and run on Wayland, but w/e. 

But I'm not understanding how the rest of the `flatpak_run_add_wayland_args` function is called if the compositor doesn't support the protocol. Is the rest of the function that falls back to that behavior done in `flatpak_run_add_wayland_security_context_args`? If so, would be nice to document that.

The confusion comes from the variable being named `security_context_available` - the final `if` check on it makes it *sound* like it's "if security context is available, fail to set up Wayland", when in reality it's just returning the initial value of the variable, `FALSE`. 

It would be nice if the variable could be set to `TRUE` initially instead, the security context setup has an `else` case that sets it to `FALSE`, and *then* the final check goes to see if the variable is *not* true, and returns `FALSE` in that case. That would make it flow a bit nicer. 

Also documenting the functions, their flow, and their behavior would be nice, to make it so conversations like this don't crop up again.

Reply:
Sorry, I don't have plans to address your comments. I think the current logic is fine, especially with the comment.

Reply:
The comment does help, and at most, I'd make it a tiny bit more explicit in the event the compositor doesn't support it:

```c
 /* If security-context is available but we failed to set it up, bail out.
  * If the compositor doesn't support security-context, we'll just
  * set up a plain Wayland socket.
  */
 ```

Reply:
That's just repeating what the code already say.

It's well known convention in flatpak that returning true means "success", so if something say

```
if (try_security_context(&was_it_even_there))
  return true;
if (was_it_even_there)
  return false;
```

is very clear. The only question is "why", not "what", which could just as well be in the commit message.